src/python/arcor2_arserver/lock/lock.py

```diff
@@ -2,112 +2,27 @@
 
 import asyncio
 from contextlib import asynccontextmanager
-from typing import AsyncGenerator, Dict, List, Optional, Tuple, Union
+from typing import AsyncGenerator, Dict, List, Optional, Set, Tuple, TypedDict, Union
+
+from websockets.server import WebSocketServerProtocol as WsClient
 
 from arcor2.cached import UpdateableCachedProject, UpdateableCachedScene
-from arcor2.data.common import BareAction, LogicItem, NamedOrientation, Project, Scene, SceneObject, StrEnum
+from arcor2.data.common import LogicItem, Project, Scene, StrEnum
 from arcor2.exceptions import Arcor2Exception
 from arcor2_arserver.clients import persistent_storage as storage
+from arcor2_arserver.decorators import retry
+from arcor2_arserver.lock.exceptions import CannotLock, LockingException
+from arcor2_arserver.lock.structures import LockedObject
 
 
-class LockingException(Arcor2Exception):
-    pass
-
-
-class CannotUnlock(LockingException):
-    pass
-
-
-class CannotLock(LockingException):
-    pass
-
-
-class Data:
-    def __init__(self, owner: str) -> None:
-        self.owners: List[str] = [owner]
-        self.count: int = 1
-
-    def inc_count(self) -> None:
-        self.count += 1
-
-    def dec_count(self) -> None:
-        self.count -= 1
-
-
-class LockedObject:
-    __slots__ = "read", "write", "tree"
-
-    def __init__(self) -> None:
-
-        # object id: data
-        self.read: Dict[str, Data] = {}
-        self.write: Dict[str, Data] = {}
-
-        self.tree: bool = False
-
-    def read_lock(self, obj_id: str, owner: str) -> bool:
-
-        if self.tree:
-            return False
-
-        if obj_id in self.write:
-            return False
-
-        already_locked = obj_id in self.read
-        if already_locked:
-            self.read[obj_id].owners.append(owner)
-            self.read[obj_id].inc_count()
-        else:
-            self.read[obj_id] = Data(owner)
-        return True
-
-    def read_unlock(self, obj_id: str, owner: str) -> None:
-
-        if obj_id not in self.read:
-            raise CannotUnlock(f"Object is not read locked by '{owner}'")
+class _QueueItem(TypedDict):
```

Comment:
Proč zrovna TypedDict? Nebyla by lepší NamedTuple, nebo dataclass? Někde dál

Reply:
Já jsem to asi v životě nepoužil, tak moc nevím jaké to má výhody / nevýhody :) Ale moc se mi nelíbí ten přístup k "propertám" přes klíč (string), když bych k tomu mohl přistupovat jako k normální property - což je takový bezpečnější / dá se to líp zkontrolovat.

Reply:
TypedDict s total=True vyžaduje všechny klíče, bez total jsou optional. Mám to tedy předělat na objekt?

Reply:
Myslím že by bylo lepší to mít jako něco co není dictionary, díky...